### DISCUSSION TITLE
[Feature Request] Analyse URL dynamic imports in providence-analytics

### DISCUSSION THREAD

**Discussion by sposh** [ing-bank/lion#1521] [Ideas]
Already commented to @tlouisse - would be great if providence-analytics could follow URL dynamic imports as well as node_modules & bower_components. For example, if you have undeclared dependencies which come from a CDN, something like this:

```js
import importLazy from 'my.libs.import-lazy';
[...]
  async importComponent() {
    if (customElements.get(this.model.getComponentName())) return;
    const className = this.model.getComponentNamePascalCase();
    const module = await importLazy(this.model.getComponentFilePath());
    const classConstructor = module[className] || module.default;
    customElements.define(this.model.getComponentName(), classConstructor);
  },
```
where `my.libs.import-lazy` exports something like `return async (url) => await import(url);`.

Not sure how easy this would be as `url` would generally be generated at runtime.

**Comment by tlouisse** [comment]
Hi,

Thanks for the request.
Contrary to what I thought before, adding extra search-targets is easy: `providence analyze match-imports -r 'local_cdn_packages/*'`

Adding out of the box support for `importLazy` would not be logical since it's not a platform standard, but we could allow a hook/plugin for the find-imports analyzer.

That plugin would get the traversed Babel AST as input and could look for this method call (similar to how dynamic imports are found).
The harder part here would be to statically determine what `this.model.getComponentFilePath()` returns.
I would need a broader context (as in the complete file and the location where this.model is defined) to see if that would be possible.

**Reply by sposh** [reply]
this.model.getComponentFilePath() would return a path to a CDN for example; and the application could have got the parameters needed to construct componentFilePath from runtime (eg an API). Local CDN may not exist. I think some sort of hook might be the best bet, where the dev can hardcode component/urls. Would Providence then be able to download the component for analysis? Or would it be easier to get the dev to use a local CDN?

**Reply by tlouisse** [reply]
That would be a really cool idea. 

It would probably require:
- a plugin that allows to detect cdn imports (if they are not imported like `import 'my-cdn-import'`)
- a way to loop over the results of `find-imports` and download all  cdn imports into 'local_cdn_packages/*' (the cdn output should have a package.json per repo)

Then it should be possible to run `providence analyze match-imports -r 'local_cdn_packages/*'`.
